cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -98,6 +104,46 @@ public static void validateConfig(SecureStore secureStore, SourceControlConfig s
     }
   }
 
+  /**
+   * Returns the <a href="https://git-scm.com/docs/git-hash-object">Git Hash</a>
+   * of the requested file path in the current HEAD. For symlinks, it returns the hash of
+   * the target file. This ensures that the file hash of a symlink is equal to the hash of the target file. It
+   * doesn't return the hash for directories. If the directory contains symlinks, we can't depend on the
+   * directory hash to ensure the contents are unchanged.
+   *
+   * @param path The absolute path on the filesystem that should include {@link RepositoryManager#getBasePath()} as a
+   *             prefix.
+   * @return The git file hash of the requested file.
+   * @throws IOException              when file isn't found, there are circular symlinks or other file IO errors.
+   * @throws IllegalArgumentException when the provided file isn't a regular file, eg: directory.
+   * @throws NotFoundException        when the file isn't committed to Git.
+   * @throws IllegalStateException    when {@link RepositoryManager#startUp()} isn't called before this.
+   */
+  public String getFileHash(Path path) throws IOException, NotFoundException {
+    validateInitialized();
+    // Git considers symlinks as separate files. So the hash of the symlink will be different from the target file.
+    // Resolve path in case of symlinks. This ensures the hash of the target file is returned.
+    path = path.toRealPath();
+    // We should not be comparing hashes of directories. If the directory contains symlinks, we can't depend on the
+    // directory hash to ensure the contents are unchanged.
+    if (!Files.isRegularFile(path)) {
+      throw new IllegalArgumentException(String.format("Path %s doesn't refer to a regular file.",
+                                                       path.toAbsolutePath()));
+    }
+    Path relativePath = sourceControlConfig.getLocalRepoPath().relativize(path);
+    // HEAD refers to the checked out commit.
+    ObjectId lastCommitId = git.getRepository().resolve(Constants.HEAD);
```

Comment:
Why do we need to resolve head here vs keeping track of the commit id when we did the check out? Does it mean we can pull in between when we checkout and when this method is called? But if that's the case, then is it possible that the file hash we get here doesn't match whatever file contents we're using for pipeline deployment?

In the future, we should try to add the main structure and then later add on details like file hash and validation. It is difficult to review whether these classes make sense without seeing how they will be used. For example, I can't answer any of the questions above without more context.

Reply:
This is for getting the hash of a pipeline after pulling it and after pushing it. The repository manager doesn't support any operation to get into a detached head state. For race conditions, I feel the service should handle synchronisation. I'm trying to have the Repository manager only handle the VCS responsibilities, so it is doing what we would expect the Git command line tool to do.
We could add the responsibility of file IO to the repository manager, so any file read or write would need to go through the Repository manager. Would like to hear your thoughts on this.

Reply:
I discussed offline with @samdgupi, and we will ensure each operation runner creates it's own repository manager, which will use it's own directory to clone remote. There will be one new operation runner for each API request. So there is no chance of race conditions.

The alternative would be to return a commit ID after each clone, push operation and make getFileHash accept the commit ID. Pull's could still suffer from race conditions when some other task could push/pull before someone actually reads the file. To prevent such cases, we decided to have a separate directory for each operation.

Reply:
The alternative where this takes a commitID is much better. It is always better to design the classes such that it impossible to misuse rather than rely on the caller to do things a very specific way. It is easier to write things correctly now, and it is much much easier to maintain the code correctly with future refactorings and modifications.

Reply:
To do this properly, you'll probably have to change some other things, like having explicit clone() and delete() methods instead of always cloning in startUp() and deleting in shutDown

Reply:
@arjan-bal Do we remember the reason of making RepositoryManager extend `AbstractIdleService`? Like Albert mentioned, perhaps we could just expose clone and delete? It's also easier to test without having to run a real git server

Reply:
@de-lan the change to implement AutoCloseable instead of extending a service has been merged: https://github.com/cdapio/cdap/pull/14927

Reply:
As discussed in the meeting, we need to accept a tracked commit to resolve the fileHash. @arjan-bal Could you pls make sure the git localRepoPath is also uuid prefixed in this PR? I guess it's a small fix in `SourceControlConfig#localRepoPath` in the constructor

Reply:
@albertshau clone is now its own method that returns a commit hash, getFileHash now accepts the commit hash as a parameter.

Reply:
@de-lan I made the change in repository manager's constructor instead. Each repository manager will create a sub directory to clone the remote.
From the name, it feels like a Configuration class should hold static data, plus we would add an expectation that a SourceControlConfiguration isn't reused with multiple repository managers.